// n8n-nodes-base/dist/nodes/DataTable/actions/row/insert.operation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.description = exports.FIELD = void 0;
exports.execute = execute;
exports.executeBulk = executeBulk;
const addRow_1 = require("../../common/addRow");
const utils_1 = require("../../common/utils");
exports.FIELD = 'insert';
const displayOptions = {
    show: {
        resource: ['row'],
        operation: [exports.FIELD],
    },
};
exports.description = [(0, addRow_1.makeAddRow)(exports.FIELD, displayOptions)];
async function execute(index) {
    const dataStoreProxy = await (0, utils_1.getDataTableProxyExecute)(this, index);
    const row = (0, addRow_1.getAddRow)(this, index);
    const insertedRows = await dataStoreProxy.insertRows([row]);
    return insertedRows.map((json) => ({ json }));
}
async function executeBulk(proxy) {
    const rows = this.getInputData().flatMap((_, i) => [(0, addRow_1.getAddRow)(this, i)]);
    const insertedRows = await proxy.insertRows(rows);
    return insertedRows.map((json, item) => ({ json, pairedItem: { item } }));
}
